evolver/device.py

```diff
@@ -1,81 +1,54 @@
+from collections import defaultdict
 import time
-import pydantic
-from typing import Annotated
-from evolver.hardware.interface import SensorDriver, EffectorDriver
+
+from evolver.base import BaseConfig, BaseInterface, ConfigDescriptor
+from evolver.connection.interface import Connection
+from evolver.controller.interface import Controller
+from evolver.hardware.interface import EffectorDriver, HardwareDriver, SensorDriver
 from evolver.serial import EvolverSerialUART
+from evolver.settings import settings
 from evolver.history import HistoryServer
 
 
 DEFAULT_SERIAL = EvolverSerialUART
 DEFAULT_HISTORY = HistoryServer
-# pydantics import string alone does not generate a schema, which breaks openapi
-# docs. We wrap it to set schema explicitly.
-ImportString = Annotated[
-    pydantic.ImportString, pydantic.WithJsonSchema({'type': 'string', 'description': 'fully qualified class name'})
-]
-
-
-class ControllerDescriptor(pydantic.BaseModel):
-    driver: ImportString
-    config: dict = {}
-
-    def driver_from_descriptor(self, evolver):
-        conf = self.driver.Config.model_validate(self.config)
-        return self.driver(evolver, conf)
-
-
-class HardwareDriverDescriptor(ControllerDescriptor):
-    calibrator: ControllerDescriptor = None
```

Comment:
Descriptors need to be _just_ ``classinfo`` and a ``config`` where everything that is configurable on a class/config should be specified and contained within the config. I.e., ``calibrator`` gets absorbed into ``config``.